javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll

```diff
@@ -0,0 +1,161 @@
+ * Provides classes for reasoning about cookies added to response without the 'secure' flag being set.
+ * A cookie without the 'secure' flag being set can be intercepted and read by a malicious user.
+ */
+
+import javascript
+
+module InsecureCookie {
+  /**
+   * `secure` property of the cookie options.
+   */
+  string flag() { result = "secure" }
+
+  /**
+   * Abstract class to represent different cases of insecure cookie settings.
+   */
+  abstract class InsecureCookies extends DataFlow::Node {
+    /**
+     * The name of the middleware/library used to set the cookie.
+     */
+    abstract string getKind();
+
+    /**
+     * The `cookie` options.
+     */
+    abstract DataFlow::Node getCookieOptionsArgument();
+
+    /**
+     * Predicate that determines if a cookie is insecure.
+     */
+    abstract predicate isInsecure();
+  }
+
+  /**
+   * A cookie set using the `express` module `cookie-session` (https://github.com/expressjs/cookie-session).
+   * The flag `secure` is set to `false` by default for HTTP, `true` by default for HTTPS (https://github.com/expressjs/cookie-session#cookie-options).
+   */
+  class InsecureCookieSession extends ExpressLibraries::CookieSession::MiddlewareInstance,
+    InsecureCookies {
+    InsecureCookieSession() { this instanceof ExpressLibraries::CookieSession::MiddlewareInstance }
+
+    override string getKind() { result = "cookie-session" }
+
+    override DataFlow::SourceNode getCookieOptionsArgument() {
+      result = this.getOption("cookie").(DataFlow::SourceNode)
+    }
+
+    DataFlow::Node getCookieFlagValue(string flag) {
+      result = this.getCookieOptionsArgument().getAPropertyWrite(flag).getRhs()
+    }
+
+    // A cookie is insecure if the `secure` flag is explicitly set to `false`.
+    override predicate isInsecure() { getCookieFlagValue(flag()).mayHaveBooleanValue(false) }
+  }
+
+  /**
+   * A cookie set using the `express` module `express-session` (https://github.com/expressjs/session).
+   * The flag `secure` is not set by default (https://github.com/expressjs/session#cookiesecure).
+   * The default value for cookie options is { path: '/', httpOnly: true, secure: false, maxAge: null }.
+   */
+  class InsecureExpressSessionCookie extends ExpressLibraries::ExpressSession::MiddlewareInstance,
+    InsecureCookies {
+    override string getKind() { result = "express-session" }
+
+    override DataFlow::SourceNode getCookieOptionsArgument() {
+      result = this.getOption("cookie").(DataFlow::SourceNode)
+    }
+
+    DataFlow::Node getCookieFlagValue(string flag) {
+      result = this.getCookieOptionsArgument().getAPropertyWrite(flag).getRhs()
+    }
+
+    // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
+    override predicate isInsecure() {
+      not exists(DataFlow::SourceNode cookieOptions |
+        cookieOptions = this.getCookieOptionsArgument() and
+        getCookieFlagValue(flag()).mayHaveBooleanValue(true)
+      ) and
```

Comment:
```suggestion
      not getCookieFlagValue(flag()).mayHaveBooleanValue(true) and
```
`getCookieFlagValue(flag)` implies the existence of `getCookieOptionsArgument()`, so this exists can be simplified.